---
title: "Analysis ecosystems for single cell biology"
subtitle: "CDNM Fellows May 2024"
author: "Vince Carey"
format: 
  revealjs:
    margin: -0.01
    mainfont: "Times New Roman"
    smaller: true
    scrollable: true
---

## Road map {.smaller}

- defining "ecosystem"
- views of a benchmarking system
- characteristic experiments in single-cell RNA-seq
- examples
    - reference-based classification
    - inference on aspects of gene expression

## Ecosystem formalism

![](eco.png)

## Benchmarking task (PMID [34949812](https://pubmed.gov/34949812)) for integrating atlas components {.smaller}

![](TheisBenchRNA.png)

## Benchmarking design

![](benchDes.png)

## Upshots

- working from single-cell atlases may become common approach for
research in genome biology; artifact removal and assurance of
validity are important components
- method evaluation is empirical
- metrics are ad hoc
- additional questions
    - Did a given single-cell experiment achieve its scientific aim?
    - What are the implications for experimental design?
        - for potential atlas contributors, and for users -- "analysis design"?
    - Is biological variation _in a human population_ accurately measurable with current designs/atlases?
        - accurate for what purpose?

## scRNAseq package snapshot

```{r lkdr,echo=FALSE}
source("process_bib.R")
datasetsPage()
```
## Using it

:::: {.columns}

::: {.column width="52%"}
```{r lkz, echo=TRUE, cache=TRUE}
library(scRNAseq)
zs = ZilionisLungData()
zs
```
:::

::: {.column width="48%"}
![](SEschema.png)
:::

::::

## SingleCellExperiment class

![](SCEschema.png)

## Simple exploration

:::: {.columns}

::: {.column width="40%"}
```{r domore,echo=TRUE,message=FALSE}
library(scuttle)
library(scater)
bad = which(
  colSums(assay(zs))==0
  )
zs = zs[,-bad]
zs = logNormCounts(zs)
# irlba:
zs = runPCA(zs)
zs2 = zs[,-which(
   is.na(zs$`Major cell type`))]
zs2
```
:::

::: {.column width="60%"}

```{r lkpl}
plotPCA(zs2,colour_by="Major cell type")
```

:::

::::

## More PCs

```{r lkpl2, echo=TRUE}
plotPCA(zs2, colour_by="Major cell type", 
   ncomponents=4, point_size=.5)
```

## Patient effect

```{r lkpl3, echo=TRUE}
plotPCA(zs2, colour_by="Patient", 
   ncomponents=4, point_size=.5)
```

## Drilling down on blood cells

```{r subset}
zblood = zs2[, which(zs2$Tissue == "blood")]
zblood
table(zblood$`Major cell type`)
```

## Interactive visualization of some PCs

```{r getpl,echo=FALSE,message=FALSE}
library(plotly)
```
```{r lkviz}
pcs = reducedDim(zblood, "PCA")
mydf = data.frame(pcs, type=zblood$`Major cell type`)
plot_ly(mydf, x=~PC2, y=~PC4, color=~type, text=~type)
```

## Upshots

- SingleCellExperiment is a compact container for
multisample sc-RNA-seq, etc.
- arbitrary metadata for features and samples
- efficient representation of assay outputs
```{r lksp,echo=TRUE}
assay(zblood[1:4,1:5])
assay(zblood["GAPDH",1:5])
```
- simple array-oriented syntax; tidy transcriptomics also

## Next steps

- Zilionis data were obtained with labels already assigned:
"Major cell types were defined by a Bayesian cell classifier with bulk whole-transcriptome profiles of FACS-sorted cell populations."

- We'll use [SingleR](https://bioconductor.org/books/release/SingleRBook/introduction.html#method-description) to reclassify, using a reference called HumanPrimaryCellAtlas

- [postponed] We'll use a new algorithm for differential expression inference to
compare expression patterns of T-cells in blood and tumor for 
a given patient

- We'll have a look at aspects of cellxgene.census in a workshop instance

## Relabeling -- scalability {.larger}

```
> hpca = celldex::HumanPrimaryCellAtlasData()
> library(BiocParallel)
> myp = MulticoreParam(4)
> system.time(s1 <- SingleR(zblood, hpca, 
     hpca$label.main, BPPARAM=myp))
   user  system elapsed 
 94.173   0.030  24.908 
> system.time(s1 <- SingleR(zblood, hpca, 
     hpca$label.main, BPPARAM=SerialParam()))
   user  system elapsed 
 72.842   0.064  72.955 
```

## Relabeling -- outcome

Rows are SingleR, columns are Zilionis "Bayesian"

![](relabeling.png)

## Conclusions

- OSCA and SingleR books are useful on-ramps to single-cell
analysis in R
- workshop.bioconductor.org has various learning materials
- standard data "containers" simplify filtering and analyzing
single cell experiments
- cellxgene.census was illustrated in the workshop document
- approaches to integrative processing of atlas elements can be benchmarked
with tooling at https://github.com/theislab/scib but recommendations are complex
